packages/studio-base/src/panels/ThreeDeeRender/renderables/FoxgloveGrid.ts

```
@@ -0,0 +1,523 @@
+      void main() {
+        vUv = uv;
+        gl_Position = projectionMatrix * modelViewMatrix * vec4(position, 1.0);
+      }
+    `,
+    fragmentShader: /* glsl */ `
+      uniform sampler2D map;
+      uniform vec4 objectId;
+      varying vec2 vUv;
+      void main() {
+        vec4 color = texture2D(map, vUv);
+        if (color.a == 0.0) {
+          discard;
+        }
+        gl_FragColor = objectId;
+      }
+    `,
+    side: THREE.DoubleSide,
+    uniforms: { map: { value: texture }, objectId: { value: [NaN, NaN, NaN, NaN] } },
+  });
+}
+
+function normalizePackedElementField(
+  field: PartialMessage<PackedElementField> | undefined,
+): PackedElementField {
+  return {
+    name: field?.name ?? "",
+    offset: field?.offset ?? 0,
+    type: field?.type ?? 0,
+  };
+}
+
+function normalizeFoxgloveGrid(message: PartialMessage<Grid>): Grid {
+  return {
+    timestamp: normalizeTime(message.timestamp),
+    pose: normalizePose(message.pose),
+    frame_id: message.frame_id ?? "",
+    row_stride: message.row_stride ?? 0,
+    cell_stride: message.cell_stride ?? 0,
+    column_count: message.column_count ?? 0,
+    cell_size: {
+      x: message.cell_size?.x ?? 1,
+      y: message.cell_size?.y ?? 1,
+    },
+    fields: message.fields?.map(normalizePackedElementField) ?? [],
+    data: normalizeByteArray(message.data),
+  };
+}
+
+function minMaxColorValues(
+  output: THREE.Vector2Tuple,
+  settings: LayerSettingsFoxgloveGrid,
+  numericType: NumericType,
+): void {
+  if (SKIP_MIN_MAX.includes(settings.colorMode)) {
+    return;
+  }
+
+  const [numericMin, numericMax] = NumericTypeMinMaxValueMap[numericType];
+  const minColorValue = settings.minValue ?? numericMin;
+  const maxColorValue = settings.maxValue ?? numericMax;
+  output[0] = minColorValue;
+  output[1] = maxColorValue;
+}
+
+const NumericTypeMinMaxValueMap: Record<NumericType, [number, number]> = {
+  [NumericType.UNKNOWN]: [Number.NEGATIVE_INFINITY, Number.POSITIVE_INFINITY],
```

Does this ever actually get used? It seems like -∞ to ∞ would never produce any useful color values. Maybe just go with `[0, 1]` for unknown or omit it if it never gets used?

---

hmm yeah, I'll go with 0, 1 for now, but If we want to omit it in the future then we'll probably want to throw an error for the unknown field. Not sure when the numericType would be unknown from the data creation perspective.